syntax = "proto3";
package ListeningParty;

import "protobuf/common.proto";
import "protobuf/track.proto";
import "protobuf/playback.proto";
import "google/protobuf/timestamp.proto";
import "google/protobuf/empty.proto";

message Room {
    string id = 1;
    string name = 2;
    string description = 3;
    string room_code = 4;
    RoomStatus status = 5;
    RoomVisibility visibility = 6;
    bool invite_only = 7;
    repeated RoomMember members = 8;
    string host_user_id = 9;
    PlaybackState playback_state = 10;
    Track current_track = 11;
    int64 current_position_ms = 12;
    int64 last_position_update = 13;
    int64 track_start_time = 14;
    bool is_playing = 15;
}

message CreateRoomRequest {
    string name = 1;
    string description = 2;
    string room_code = 3;
    RoomVisibility visibility = 4;
    bool invite_only = 5;
    string host_user_id = 6;
}

message GetRoomRequest {
    string room_id = 1;
}

message GetRoomByCodeRequest {
    string room_code = 1;
}

message UpdateRoomRequest {
    string room_id = 1;
    optional string name = 2;
    optional string description = 3;
    optional RoomVisibility visibility = 4;
    optional bool invite_only = 5;
}

message DeleteRoomRequest {
    string room_id = 1;
}

message ListRoomsRequest {
    optional RoomVisibility visibility_filter = 1;
    optional string host_user_id_filter = 2;
    optional bool include_private_rooms_if_member = 3;
    int32 page_size = 4;
    string page_token = 5;
}

message ListRoomsResponse {
    repeated Room rooms = 1;
    string next_page_token = 2;
}

message JoinRoomRequest {
    oneof identifier {
        string room_id = 1;
        string room_code = 2;
    }
    string user_id = 3;
}

message LeaveRoomRequest {
    string room_id = 1;
    string user_id = 2;
}

message InviteToRoomRequest {
    string room_id = 1;
    string inviter_user_id = 2;
    string invitee_user_id = 3;
}

message KickFromRoomRequest {
    string room_id = 1;
    string performing_user_id = 2;
    string kicked_user_id = 3;
}

message RoomMember {
    string room_id = 1;
    string user_id = 2;
    RoomMemberRole role = 3;
    optional string invited_by = 4;
    google.protobuf.Timestamp joined_at = 5;
    optional google.protobuf.Timestamp left_at = 6;
}

service RoomService {
    rpc CreateRoom(CreateRoomRequest) returns (Room);
    rpc GetRoom(GetRoomRequest) returns (Room);
    rpc UpdateRoom(UpdateRoomRequest) returns (Room);
    rpc DeleteRoom(DeleteRoomRequest) returns (google.protobuf.Empty);
    rpc ListRooms(ListRoomsRequest) returns (ListRoomsResponse);
    rpc JoinRoom(JoinRoomRequest) returns (Room);
    rpc LeaveRoom(LeaveRoomRequest) returns (google.protobuf.Empty);
    rpc InviteToRoom(InviteToRoomRequest) returns (google.protobuf.Empty);
    rpc KickFromRoom(KickFromRoomRequest) returns (google.protobuf.Empty);
}
